### PR TITLE
add transformed mvn autoguide

### PR DESCRIPTION
**Pull request by jpchen**
the simplest of fancier autoguides:
 - fixes bugs in auto reg nn
 - implements a simple autoguide learns gaussian noise generated via iaf
 - increases num iters to pass cuda tests (#1177)

### CODE REVIEW

File: pyro/contrib/autoguide/__init__.py [hunk 1]
```diff
@@ -519,6 +520,52 @@ def _loc_scale(self, *args, **kwargs):
         return loc, scale
 
 
+class AutoTransformedNormal(AutoContinuous):
```
Comment (jpchen) [inline pyro/contrib/autoguide/__init__.py]: open to more descriptive names that arent egregiously long

File: pyro/contrib/autoguide/__init__.py [hunk 2]
```diff
@@ -519,6 +520,52 @@ def _loc_scale(self, *args, **kwargs):
...
+        Samples the (single) multivariate normal latent used in the auto guide.
+        """
+        loc = pyro.param("{}_loc".format(self.prefix),
+                         lambda: torch.zeros(self.latent_dim))
+        W_term = pyro.param("{}_W_term".format(self.prefix),
+                            lambda: torch.randn(self.rank, self.latent_dim) * (0.5 / self.rank) ** 0.5)
+        D_term = pyro.param("{}_D_term".format(self.prefix),
+                            lambda: torch.ones(self.latent_dim) * 0.5,
+                            constraint=constraints.positive)
+        iaf = dist.InverseAutoregressiveFlow(self.latent_dim, 10*self.latent_dim)
```
Comment (jpchen) [inline pyro/contrib/autoguide/__init__.py]: `10*self.latent_dim` is a bit ad-hoc... suggestions?

File: pyro/contrib/autoguide/__init__.py [hunk 3]
```diff
@@ -519,6 +520,42 @@ def _loc_scale(self, *args, **kwargs):
         return loc, scale
 
 
+class AutoTransformedNormal(AutoLowRankMultivariateNormal):
```
Comment (jpchen) [inline pyro/contrib/autoguide/__init__.py]: open to more descriptive names that arent egregiously long

Comment (fritzo) [inline pyro/contrib/autoguide/__init__.py]: You should not inherit from `AutoLowRankMultivariateNormal`, because that class assumes the `._loc_scale()` method produces quantiles of the models variables (not quantiles of an arbitrary latent). Instead I would recommend forking that class's code and defining custom versions of its `.sample_latent()` and `._loc_scale()` methods so that `.median()` and `.quantiles()` will be defined correctly.

File: tests/contrib/autoguide/test_advi.py [hunk 4]
```diff
@@ -113,15 +118,25 @@ def model():
         infer.step()
 
     median = guide.median()
-    assert_equal(median["x"], torch.tensor(0.0), prec=0.1)
-    if auto_class is AutoDelta:
-        assert_equal(median["y"], torch.tensor(-1.0).exp(), prec=0.1)
+    if auto_class is AutoTransformedNormal:
```
[the remark this answers is not in the document]
Comment (jpchen) [reply tests/contrib/autoguide/test_advi.py]: yes for 1e3 steps and 1e-3 lr, ill modify the hyperparams instead since it's only for this one class

File: pyro/contrib/autoguide/__init__.py [hunk 5]
```diff
@@ -561,6 +562,57 @@ def _loc_scale(self, *args, **kwargs):
...
+        self.iaf_dist = dist.TransformedDistribution(dist.LowRankMultivariateNormal(loc, W_term, D_term), [iaf])
+        return pyro.sample("_{}_latent".format(self.prefix), self.iaf_dist, infer={"is_auxiliary": True})
+
+    def _loc_scale(self, *args, **kwargs):
+        loc = pyro.param("{}_loc".format(self.prefix))
+        W_term = pyro.param("{}_W_term".format(self.prefix))
+        D_term = pyro.param("{}_D_term".format(self.prefix))
+        scale = (W_term.pow(2).sum(0) + D_term).sqrt()
+        transform = biject_to(self.iaf_dist.support)
+        loc, scale = transform(loc), transform(scale)
```
Comment (fritzo) [inline pyro/contrib/autoguide/__init__.py]: nonsense

File: pyro/contrib/autoguide/__init__.py [hunk 6]
```diff
@@ -561,6 +562,57 @@ def _loc_scale(self, *args, **kwargs):
...
+                         lambda: torch.zeros(self.latent_dim))
+        W_term = pyro.param("{}_W_term".format(self.prefix),
+                            lambda: torch.randn(self.rank, self.latent_dim) * (0.5 / self.rank) ** 0.5)
+        D_term = pyro.param("{}_D_term".format(self.prefix),
+                            lambda: torch.ones(self.latent_dim) * 0.5,
+                            constraint=constraints.positive)
+        hidden_dim = self.hidden_dim if self.hidden_dim is not None else self.latent_dim
+        iaf = dist.InverseAutoregressiveFlow(self.latent_dim, hidden_dim, sigmoid_bias=self.sigmoid_bias)
+        pyro.module("{}_iaf".format(self.prefix), iaf.module)
+        self.iaf_dist = dist.TransformedDistribution(dist.LowRankMultivariateNormal(loc, W_term, D_term), [iaf])
```
Comment (fritzo) [inline pyro/contrib/autoguide/__init__.py]: I think we instead want to transform white noise via the iaf, then apply learnable diagonal noise:
```py
def sample_latent(self, *args, **kwargs):
    iaf = self._get_iaf()  # calls pyro.module internally
    iaf_dist = dist.TransformedDistribution(dist.Normal(0., 1.).expand([self.hidden_dim]), iaf)
    loc = pyro.sample("_{}_loc".format(self.prefix), iaf_dist.independent(1))
    scale = pyro.param("{}_scale".format(self.prefix), lambda: torch.ones(self.latent_dim))
    return pyro.sample("_{}_latent".format(self.prefix), dist.Normal(loc, scale).independent(1))

def _loc_scale(self, *args, **kwargs):
    iaf = self._get_iaf()  # calls pyro.module internally
    prototype = pyro.param("{}_scale".format(self.prefix))
    loc = iaf(prototype.new_zeros(self.hidden_dim))
    scale = None  # no analytical solution
    return loc, scale
```

File: pyro/contrib/autoguide/__init__.py [hunk 7]
```diff
@@ -417,6 +418,8 @@ def quantiles(self, quantiles, *args, **kwargs):
         :rtype: dict
         """
         loc, scale = self._loc_scale(*args, **kwargs)
+        if loc is None or scale is None:
+            raise ValueError('loc and scale cannot be None')
```
Comment (fritzo) [inline pyro/contrib/autoguide/__init__.py]: Thanks for being safe. How about:
```py
if scale is None:
    raise NotImplementedError('scale cannot be computed analytically')
```

File: pyro/contrib/autoguide/__init__.py [hunk 8]
```diff
@@ -561,6 +564,51 @@ def _loc_scale(self, *args, **kwargs):
...
+    over the entire latent space. The guide does not depend on the model's
+    ``*args, **kwargs``.
+
+    Usage::
+
+        guide = AutoTransformedNormal(model, rank=10)
+        svi = SVI(model, guide, ...)
+
+    :param callable model: a generative model
+    :param int rank: the rank of the low-rank part of the covariance matrix
```
Comment (fritzo) [inline pyro/contrib/autoguide/__init__.py]: You can now omit `rank` throughout.

File: pyro/contrib/autoguide/__init__.py [hunk 9]
```diff
@@ -561,6 +564,51 @@ def _loc_scale(self, *args, **kwargs):
         return loc, scale
 
 
+class AutoTransformedNormal(AutoContinuous):
+    """
+    This implementation of :class:`AutoContinuous` uses a Transformed
+    Low Rank MV Normal distribution via a IAF to construct a guide
```
Comment (fritzo) [inline pyro/contrib/autoguide/__init__.py]: Low Rank MV Normal -> Diagonal Normal

File: pyro/nn/auto_reg_nn.py [hunk 10]
```diff
@@ -65,7 +65,7 @@ def __init__(self, input_dim, hidden_dim, output_dim_multiplier=1,
 
         if mask_encoding is None:
             # the dependency structure is chosen at random
-            self.mask_encoding = 1 + torch.multinomial(torch.ones(input_dim - 1) / (input_dim - 1),
+            self.mask_encoding = 1 + torch.multinomial(torch.tensor(input_dim - 1).fill_(1.).float(),
```
Comment (fritzo) [inline pyro/nn/auto_reg_nn.py]: The input probs are no longer normalized. Can you confirm that `torch.multinomial` normalizes internally, so that your change is valid?

File: pyro/contrib/autoguide/__init__.py [hunk 11]
```diff
@@ -561,6 +564,51 @@ def _loc_scale(self, *args, **kwargs):
...
+        self.sigmoid_bias = sigmoid_bias
+        self.hidden_dim = hidden_dim
+        self.rank = rank
+        super(AutoTransformedNormal, self).__init__(model, prefix)
+
+    def sample_latent(self, *args, **kwargs):
+        """
+        Samples the (single) multivariate normal latent used in the auto guide.
+        """
+        hidden_dim = self.hidden_dim if self.hidden_dim is not None else self.latent_dim
```
Comment (fritzo) [inline pyro/contrib/autoguide/__init__.py]: In might be simpler to
```py
if self.hidden_dim is None:
    self.hidden_dim = self.latent_dim
```
and thereafter refer only to `self.hidden_dim`.

File: pyro/contrib/autoguide/__init__.py (hunk 7 above)
Comment (jpchen) [reply pyro/contrib/autoguide/__init__.py]: this is in the `AutoContinuous` class so should error if either `loc` or `scale` isnt computed with a general error message. do you want an overridden check in `AutoTransformedNormal`?

File: pyro/nn/auto_reg_nn.py (hunk 10 above)
Comment (jpchen) [reply pyro/nn/auto_reg_nn.py]: see [note](https://pytorch.org/docs/stable/torch.html?highlight=multinomial#torch.multinomial)

File: pyro/contrib/autoguide/__init__.py (hunk 11 above)
Comment (jpchen) [reply pyro/contrib/autoguide/__init__.py]: yeah i did that at first then didnt want masked errors if `_loc_scale` is called before `sample_latent` since `self.latent_dim` is only constructed during `__call__`

Comment (fritzo) [reply pyro/contrib/autoguide/__init__.py]: I think we should just error if `_loc_scale()` is called before `__call__()`

File: pyro/contrib/autoguide/__init__.py (hunk 7 above)
Comment (fritzo) [reply pyro/contrib/autoguide/__init__.py]: I just think we should raise a `NotImplementedError` rather than a `ValueError`. Feel free to `raise NotImplementedError` with an empty error message.

**Comment by martinjankowiak** [comment]
let me take a look at this please

**Comment by jpchen** [comment]
@fritzo im happy to merge as is (after martin takes a look), per our discussions offline, i may want to add some additional bells and whistles after running them on the example suite

**Comment by jpchen** [comment]
@martinjankowiak @fritzo please take another look.  i think this is what we converged to, per our discussion offline. adding multiple iafs/nns can be done in a follow up

File: pyro/contrib/autoguide/__init__.py [hunk 12]
```diff
@@ -561,6 +562,49 @@ def _loc_scale(self, *args, **kwargs):
         return loc, scale
 
 
+class AutoTransformedNormal(AutoContinuous):
+    """
+    This implementation of :class:`AutoContinuous` uses a Transformed
+    Diagonal Normal distribution via an IAF to construct a guide
```
Comment (martinjankowiak) [inline pyro/contrib/autoguide/__init__.py]: uses a Diagonal Normal distribution transformed by an Inverse Autoregressive Flow

Comment (martinjankowiak) [reply pyro/contrib/autoguide/__init__.py]: also it should probably just becalled something like AutoIAFNormal

File: pyro/contrib/autoguide/__init__.py [hunk 13]
```diff
@@ -561,6 +562,49 @@ def _loc_scale(self, *args, **kwargs):
...
+    """
+    def __init__(self, model, hidden_dim=None, sigmoid_bias=2.0, prefix="auto"):
+        self.sigmoid_bias = sigmoid_bias
+        self.hidden_dim = hidden_dim
+        super(AutoTransformedNormal, self).__init__(model, prefix)
+
+    def sample_latent(self, *args, **kwargs):
+        if self.hidden_dim is None:
+            self.hidden_dim = self.latent_dim
+        loc = pyro.param("{}_loc".format(self.prefix),
```
Comment (martinjankowiak) [inline pyro/contrib/autoguide/__init__.py]: loc/param shouldn't be trainable

File: pyro/contrib/autoguide/__init__.py [hunk 14]
```diff
@@ -561,6 +562,49 @@ def _loc_scale(self, *args, **kwargs):
...
+                           constraint=constraints.positive)
+        iaf = dist.InverseAutoregressiveFlow(self.latent_dim, self.hidden_dim,
+                                             sigmoid_bias=self.sigmoid_bias)
+        pyro.module("{}_iaf".format(self.prefix), iaf.module)
+        self.iaf_dist = dist.TransformedDistribution(dist.Normal(loc, scale), [iaf])
+        return pyro.sample("_{}_latent".format(self.prefix), self.iaf_dist.independent(1),
+                           infer={"is_auxiliary": True})
+
+    def _loc_scale(self, *args, **kwargs):
+        loc = pyro.param("{}_loc".format(self.prefix))
```
Comment (martinjankowiak) [inline pyro/contrib/autoguide/__init__.py]: if _loc_scale is used for median this doesn't make sense

File: pyro/nn/auto_reg_nn.py [hunk 15]
```diff
@@ -83,8 +83,11 @@ def __init__(self, input_dim, hidden_dim, output_dim_multiplier=1,
         self.mask2 = torch.zeros(input_dim * self.output_dim_multiplier, hidden_dim)
 
         for k in range(hidden_dim):
-            # fill in mask1
-            m_k = self.mask_encoding[k].item()
+            if self.mask_encoding.dim():
```
Comment (martinjankowiak) [inline pyro/nn/auto_reg_nn.py]: why is this needed?

File: pyro/contrib/autoguide/__init__.py (hunk 14 above)
Comment (fritzo) [reply pyro/contrib/autoguide/__init__.py]: Right, you should simply omit this method

File: pyro/nn/auto_reg_nn.py (hunk 15 above)
Comment (jpchen) [reply pyro/nn/auto_reg_nn.py]: you cant index a scalar.

File: pyro/contrib/autoguide/__init__.py [hunk 16]
```diff
@@ -561,6 +562,39 @@ def _loc_scale(self, *args, **kwargs):
         return loc, scale
 
 
+class AutoIAFNormal(AutoContinuous):
+    """
+    This implementation of :class:`AutoContinuous` uses a Transformed
```
Comment (martinjankowiak) [inline pyro/contrib/autoguide/__init__.py]: uses a Diagonal Normal distribution transformed via an IAF

File: pyro/contrib/autoguide/__init__.py [hunk 17]
```diff
@@ -561,6 +562,39 @@ def _loc_scale(self, *args, **kwargs):
...
+    Diagonal Normal distribution via an IAF to construct a guide
+    over the entire latent space. The guide does not depend on the model's
+    ``*args, **kwargs``.
+
+    Usage::
+
+        guide = AutoIAFNormal(model, rank=10)
```
Comment (martinjankowiak) [inline pyro/contrib/autoguide/__init__.py]: rank is gone

File: tests/contrib/autoguide/test_advi.py [hunk 18]
```diff
@@ -179,6 +186,7 @@ def model():
     AutoDiagonalNormal,
     AutoMultivariateNormal,
     AutoLowRankMultivariateNormal,
+    AutoIAFNormal,
```
Comment (martinjankowiak) [inline tests/contrib/autoguide/test_advi.py]: wasn't this to be removed?

are there any "actual" tests of autoiafnormal here?

File: pyro/contrib/autoguide/__init__.py (hunk 16 above)
Comment (martinjankowiak) [reply pyro/contrib/autoguide/__init__.py]: link to IAF docs

File: pyro/nn/auto_reg_nn.py (hunk 15 above)
Comment (martinjankowiak) [reply pyro/nn/auto_reg_nn.py]: i'd remove this. cuz of an IAF is 1D it'll just be an affine transformation of the base distribution. raise an error if user wants 1d autoiaf?

Comment (jpchen) [reply pyro/nn/auto_reg_nn.py]: we can catch this for auto iaf, but like you said a 1d iaf is a valid thing, and even if it werent that error should be caught, not throw a pytorch tensor error.

File: tests/contrib/autoguide/test_advi.py (hunk 18 above)
Comment (jpchen) [reply tests/contrib/autoguide/test_advi.py]: why would this be removed?

Comment (jpchen) [reply tests/contrib/autoguide/test_advi.py]: why is that needed for testing parallel enumeration?

File: pyro/nn/auto_reg_nn.py [hunk 19]
```diff
@@ -60,12 +62,14 @@ def __init__(self, input_dim, hidden_dim, output_dim_multiplier=1,
                  mask_encoding=None, permutation=None):
         super(AutoRegressiveNN, self).__init__()
         self.input_dim = input_dim
+        if hidden_dim == 1:
+            warnings.warn('AutoRegressiveNN hidden_dim = 1. Consider using an affine transformation instead.')
```
Comment (martinjankowiak) [inline pyro/nn/auto_reg_nn.py]: this should be input_dim

**Comment by eb8680** [comment]
@martinjankowiak good to merge?
